bug: wasm build error

### Did you check existing issues?

- [X] I have read all the [tree-sitter docs](https://tree-sitter.github.io/tree-sitter/using-parsers) if it relates to using the parser
- [X] I have searched the existing issues of tree-sitter-typescript

### Tree-Sitter CLI Version, if relevant (output of `tree-sitter --version`)

tree-sitter 0.22.1 (1c38d34deac5ae7ea797a559f596fb4ec9008eb4)

### Describe the bug

scanner.c:1:10: fatal error: '../../common/scanner.h' file not found
#include "../../common/scanner.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
emcc: error: '/emsdk/upstream/bin/clang -target wasm32-unknown-emscripten -fignore-exceptions -fPIC -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -DEMSCRIPTEN -Werror=implicit-function-declaration --sysroot=/emsdk/upstream/emscripten/cache/sysroot -Xclang -iwithsysroot/include/fakesdl -Xclang -iwithsysroot/include/compat -Os -fno-exceptions -fvisibility=hidden -I. scanner.c -c -o /tmp/emscripten_temp_zb81c6j6/scanner_0.o' failed (returned 1)
emcc command failed

### Steps To Reproduce/Bad Parse Tree

I build all tree-sitter supported language repo find typescript/php/ocaml throw same error ,I'm not sure if it's the same problem

### Expected Behavior/Parse Tree

build success

### Repro

```TypeScript
// Example code that causes the issue
function foo() {
  // Code that fails to parse, or causes an error
}
```

--- a/cli/loader/src/lib.rs
+++ b/cli/loader/src/lib.rs
@@ -475,6 +475,7 @@ impl Loader {
             if recompile {
                 self.compile_parser_to_wasm(
                     &config.name,
+                    None,
                     config.src_path,
                     config
                         .scanner_path
@@ -723,6 +724,7 @@ impl Loader {
     pub fn compile_parser_to_wasm(
         &self,
         language_name: &str,
+        root_path: Option<&Path>,
         src_path: &Path,
         scanner_filename: Option<&Path>,
         output_path: &Path,
@@ -735,6 +737,7 @@ impl Loader {
             Podman,
         }
 
+        let root_path = root_path.unwrap_or(src_path);
         let emcc_name = if cfg!(windows) { "emcc.bat" } else { "emcc" };
 
         // Order of preference: emscripten > docker > podman > error
@@ -773,10 +776,18 @@ impl Loader {
                 };
                 command.args(["run", "--rm"]);
 
-                // Mount the parser directory as a volume
-                command.args(["--workdir", "/src"]);
+                // The working directory is the directory containing the parser itself
+                let workdir = if root_path == src_path {
+                    PathBuf::from("/src")
+                } else {
+                    let mut path = PathBuf::from("/src");
+                    path.push(src_path.strip_prefix(root_path).unwrap());
+                    path
+                };
+                command.args(["--workdir", &workdir.to_string_lossy()]);
 
-                let mut volume_string = OsString::from(&src_path);
+                // Mount the root directory as a volume, which is the repo root
+                let mut volume_string = OsString::from(&root_path);
                 volume_string.push(":/src:Z");
                 command.args([OsStr::new("--volume"), &volume_string]);
 

--- a/cli/src/generate/grammar_files.rs
+++ b/cli/src/generate/grammar_files.rs
@@ -60,7 +60,7 @@ const PACKAGE_SWIFT_TEMPLATE: &str = include_str!("./templates/Package.swift");
 struct LanguageConfiguration {}
 
 #[derive(Deserialize, Debug)]
-struct PackageJSON {
+pub struct PackageJSON {
     #[serde(rename = "tree-sitter")]
     tree_sitter: Option<Vec<LanguageConfiguration>>,
 }
@@ -458,7 +458,7 @@ pub fn generate_grammar_files(
     Ok(())
 }
 
-fn lookup_package_json_for_path(path: &Path) -> Result<(PathBuf, PackageJSON)> {
+pub fn lookup_package_json_for_path(path: &Path) -> Result<(PathBuf, PackageJSON)> {
     let mut pathbuf = path.to_owned();
     loop {
         let package_json = pathbuf

--- a/cli/src/generate/mod.rs
+++ b/cli/src/generate/mod.rs
@@ -28,6 +28,8 @@ mod render;
 mod rules;
 mod tables;
 
+pub use grammar_files::lookup_package_json_for_path;
+
 lazy_static! {
     static ref JSON_COMMENT_REGEX: Regex = RegexBuilder::new("^\\s*//.*")
         .multi_line(true)

--- a/cli/src/main.rs
+++ b/cli/src/main.rs
@@ -11,10 +11,11 @@ use glob::glob;
 use regex::Regex;
 use tree_sitter::{ffi, Parser, Point};
 use tree_sitter_cli::{
-    generate, highlight, logger,
+    generate::{self, lookup_package_json_for_path},
+    highlight, logger,
     parse::{self, ParseFileOptions, ParseOutput},
-    playground, query, tags, test,
-    test::TestOptions,
+    playground, query, tags,
+    test::{self, TestOptions},
     test_highlight, test_tags, util, wasm,
 };
 use tree_sitter_config::Config;
@@ -463,8 +464,11 @@ fn run() -> Result<()> {
                 let grammar_path =
                     current_dir.join(build_options.path.as_deref().unwrap_or_default());
                 let output_path = build_options.output.map(|path| current_dir.join(path));
+                let root_path = lookup_package_json_for_path(&grammar_path.join("package.json"))
+                    .map(|(p, _)| p.parent().unwrap().to_path_buf())?;
                 wasm::compile_language_to_wasm(
                     &loader,
+                    Some(&root_path),
                     &grammar_path,
                     &current_dir,
                     output_path,
@@ -515,8 +519,11 @@ fn run() -> Result<()> {
         Commands::BuildWasm(wasm_options) => {
             eprintln!("`build-wasm` is deprecated and will be removed in v0.24.0. You should use `build --wasm` instead");
             let grammar_path = current_dir.join(wasm_options.path.unwrap_or_default());
+            let root_path = lookup_package_json_for_path(&grammar_path.join("package.json"))
+                .map(|(p, _)| p.parent().unwrap().to_path_buf())?;
             wasm::compile_language_to_wasm(
                 &loader,
+                Some(&root_path),
                 &grammar_path,
                 &current_dir,
                 None,

--- a/cli/src/wasm.rs
+++ b/cli/src/wasm.rs
@@ -33,6 +33,7 @@ pub fn get_grammar_name(language_dir: &Path) -> Result<String> {
 
 pub fn compile_language_to_wasm(
     loader: &Loader,
+    root_dir: Option<&Path>,
     language_dir: &Path,
     output_dir: &Path,
     output_file: Option<PathBuf>,
@@ -45,6 +46,7 @@ pub fn compile_language_to_wasm(
     let scanner_path = loader.get_scanner_path(&src_path);
     loader.compile_parser_to_wasm(
         &grammar_name,
+        root_dir,
         &src_path,
         scanner_path
             .as_ref()

